pvlib/iotools/psm3.py

```diff
@@ -0,0 +1,148 @@
+    Get PSM3 data
+
+    Parameters
+    ----------
+    latitude : float or int
+        in decimal degrees, between -90 and 90, north is positive
+    longitude : float or int
+        in decimal degrees, between -180 and 180, east is positive
+    names : str
+        PSM3 API parameter specifing year or TMY variant to download, see notes
+        below for options, default: ``'tmy'``
+    interval : int
+        interval size in minutes, can be only either 30 or 60, default: 60
+    api_key : str
+        optional, supply your NREL Developer Network API key
+
+    Returns
+    -------
+    headers : dict
+        metadata from NREL PSM3 about the record, see notes for fields
+    data : pandas.DataFrame
+        timeseries data from NREL PSM3
+
+    Raises
+    ------
+    requests.HTTPError
+        if the request return status is not ok then the ``'errors'`` from the
+        JSON response will be returned as an exception
+
+    Notes
+    -----
+    The PSM3 API `names` parameter must be a single value from the following
+    list::
+
+        ['1998', '1999', '2000', '2001', '2002', '2003', '2004', '2005',
+         '2006', '2007', '2008', '2009', '2010', '2011', '2012', '2013',
+         '2014', '2015', '2016', '2017', 'tmy', 'tmy-2016', 'tmy-2017',
+         'tdy-2017', 'tgy-2017']
+
+    The return is a tuple with two items. The first item is a header with
+    metadata from NREL PSM3 about the record containing the following fields:
+
+    * Source
+    * Location ID
+    * City
+    * State
+    * Country
+    * Latitude
+    * Longitude
+    * Time Zone
+    * Elevation
+    * Local Time Zone
+    * Dew Point Units
+    * DHI Units
+    * DNI Units
+    * GHI Units
+    * Temperature Units
+    * Pressure Units
+    * Wind Direction Units
+    * Wind Speed
+    * Surface Albedo Units
+    * Version
+
+    The second item is a dataframe with the timeseries data downloaded.
+
+    See Also
+    --------
+    pvlib.iotools.read_tmy2, pvlib.iotools.read_tmy3
+
+    References
+    ----------
+
+    * `NREL Developer Network - Physical Solar Model (PSM) v3
+      <https://developer.nrel.gov/docs/solar/nsrdb/psm3_data_download/>`_
+    * `NREL National Solar Radiation Database (NSRDB)
+      <https://nsrdb.nrel.gov/>`_
+
+    """
+    longitude = ('%9.4f' % longitude).strip()
+    latitude = ('%8.4f' % latitude).strip()
+    params = {
+        'api_key': api_key,
+        'full_name': 'Sample User',
+        'email': 'sample@email.com',
+        'affiliation': 'Test Organization',
+        'reason': 'Example',
+        'mailing_list': 'true',
+        'wkt': 'POINT(%s %s)' % (longitude, latitude),
+        'names': names,
+        'attributes':  ','.join(ATTRIBUTES),
+        'leap_day': 'false',
+        'utc': 'false',
+        'interval': interval
+    }
+    # request CSV download from NREL PSM3
+    response = requests.get(URL, params=params)
+    if not response.ok:
+        raise requests.HTTPError(response.json()['errors'])
```

Comment:
I think it would be helpful to return the error message, or None.  In my workflow I'm looping over lat, long, and so put the call to `get_psm3` inside a `try:except` block to avoid exiting the loop on the first error.  NREL puts information in the response error when, for example, a lat/long is passed where there is no NSRDB data.  In my case I could decide what to do in the `except` block based on the error message.

To be explicit:
`return header, data, response.json()['errors']`

Reply:
I don't understand. AFAIK, that's exactly what this does. here's my workflow:
```python
try:
    header, data = get_psm3(latitude, longitude)
except requests.HTTPError as exc:
    print(exc)
```

but actually it looks more like this
```python
@app.route('/', methods=['GET', 'POST'])
def myfun(latitude=LATITUDE, longitude=LONGITUDE, tracker=False,
            surface_azimuth=0.0, surface_tilt=20.0, module_type=None,
            project_size=None, dst=False):
    kwargs = {'title': 'PVWatts'}
    if request.method == 'GET':
        try:
            # FYI: calculate() is a method that calls get_psm3()
            mpp, header = calculate(
                latitude, longitude, tracker, surface_tilt, surface_azimuth)
        except requests.HTTPError as exc:
            flash(exc, 'error')
            return render_template('pvwatts.html', title='PVWatts')
        else:
            # do stuff
```
so if I put a location outside the US, like Bristol-UK, then it flashes the errors on the screen, exactly as they were returned by psm3
![get_psm3](https://user-images.githubusercontent.com/1385621/57049174-b81fdc00-6c2b-11e9-98fd-e82170a696dc.png)
Does this work for you?

Reply:
@cwhanse also note that if the response, `r`, is successful, it returns the CSV data in `r.content`, and `r.json()` doesn't work. Therefore `r.json()['errors']` only works if the response is **not** okay. And vice-versa, if the response isn't okat, then there's no CSV file, and therefore, no header or data., So the suggestion to do this:
```python
return header, data, response.json()['errors']
```
wouldn't work, it would have to be more complicated, like this:
```python
if r.ok:
    return pvlib.iotools.read_tmy3(StringIO(response.content))
else:
    return response.json()['errors']
```
PyLint would complain about this signature because it returns different types and number of returns depending on the outcome which is not a consistent API. We could do some workarounds to make it consistent, like you say, return `None` when headers, data, or errors are missing, but IMO that's not as clean as just raising an exception and letting the caller decide what they want to do.

Reply:
My workflow issues appear to be my misunderstanding of how to use `try/except` in combination with errors raised by a function called in the `try` block.  From my point of view, `get_psm3` is enahnced if it can help with error management without compromising its purpose (get and return PSM3 data for one location). Merely displaying the error and exiting is likely to cause frustration.  But I don't have a concrete suggestion if returning the error isn't desireable.

Reply:
In general in python, if errors cannot be handled within the function, they should be raised rather than returned.

Reply:
@cwhanse [`requests.HTTPError`](https://2.python-requests.org/en/master/api/#requests.HTTPError) is a subclass of the base `RequestsException` that also has a `response` attribute that has the entire response from the request, including the `json()`, the `content`, and the `status_code` so you can catch the exception and handle it anyway you want. Printing out the message is just one option, you can get the response and do something else:
```python
import requests
URL = "http://developer.nrel.gov/api/solar/nsrdb_psm3_download.csv"
r = requests.get(URL)
x = requests.HTTPError(r.content, response=r)

# print message
try:
    raise x
except requests.HTTPError as exc:
    print(exc)
# b'Error Code,Error Message\n"API_KEY_MISSING","No api_key was supplied. Get one at https://developer.nrel.gov/signup/"'

# print response
try:
    raise x
except requests.HTTPError as exc:
    print(exc.response)
# <Response [403]>

# handle exception based on response status code
try:
    raise x
except requests.HTTPError as exc:
    if exc.response.status_code == 403:
        # do something else

x.response.__dict__
# ouptut
{'_content': b'Error Code,Error Message\n"API_KEY_MISSING","No api_key was supplied. Get one at https://developer.nrel.gov/signup/"',
 '_content_consumed': True,
 '_next': None,
 'status_code': 403,
 'headers': {'Server': 'openresty', 'Date': 'Thu, 02 May 2019 17:06:46 GMT', 'Content-Type': 'text/csv', 'Transfer-Encoding': 'chunked', 'Connection': 'keep-alive', 'Vary': 'Accept-Encoding', 'Access-Control-Allow-Origin': '*', 'X-Cache': 'MISS', 'Strict-Transport-Security': 'max-age=31536000; preload', 'Content-Encoding': 'gzip'},
 'raw': <urllib3.response.HTTPResponse at 0x15e832f92b0>,
 'url': 'http://developer.nrel.gov/api/solar/nsrdb_psm3_download.csv',
 'encoding': 'ISO-8859-1',
 'history': [],
 'reason': 'Forbidden',
 'cookies': <RequestsCookieJar[]>,
 'elapsed': datetime.timedelta(0, 0, 250634),
 'request': <PreparedRequest [GET]>,
 'connection': <requests.adapters.HTTPAdapter at 0x15e832fc2e8>}
```

Reply:
or `if 'something I know how to handle' in str(exc): handle_it()`

Reply:
Empty or rejected responses due to the throttle limit on requests throw an error when parsing the response, see the comments below.  I suspect this situation will come up.  As far as I can tell, in this situation the status_code is 400 but the errors text is empty.